File: typescript/agent.ts
/**
 * Payment service agent — TypeScript example.
 *
 * Processes payments and resumes with transaction result.
 *
 * Usage:
 *   export AXME_API_KEY="<agent-key>"
 *   npx tsx agent.ts
 */

import { AxmeClient } from "@axme/axme";

const AGENT_ADDRESS = "payment-service-demo";

async function handleIntent(client: AxmeClient, intentId: string) {
  const intentData = await client.getIntent(intentId);
  const intent = intentData.intent ?? intentData;
  let payload = intent.payload ?? {};
  if (payload.parent_payload) {
    payload = payload.parent_payload;
  }

  const orderId = payload.order_id ?? "unknown";
  const amount = payload.amount ?? 0;
  const currency = payload.currency ?? "USD";
  const method = payload.method ?? "card";

  console.log(`  Processing ${method} payment: ${currency} ${amount} for ${orderId}...`);
  await new Promise((r) => setTimeout(r, 1000));
  console.log(`  Authorizing with payment provider...`);
  await new Promise((r) => setTimeout(r, 1000));
  console.log(`  Capturing funds...`);
  await new Promise((r) => setTimeout(r, 1000));

  const result = {
    action: "complete",
    order_id: orderId,
    transaction_id: "TXN-99001",
    amount_charged: amount,
    status: "captured",
    processed_at: new Date().toISOString(),
  };

  await client.resumeIntent(intentId, result, { ownerAgent: "payment-service-demo" });
  console.log(`  Payment captured: TXN-99001 (${currency} ${amount})`);
}

async function main() {
  const apiKey = process.env.AXME_API_KEY;
  if (!apiKey) {
    console.error("Error: AXME_API_KEY not set.");
    process.exit(1);
  }

  const client = new AxmeClient({ apiKey });

  console.log(`Agent listening on ${AGENT_ADDRESS}...`);
  console.log("Waiting for intents (Ctrl+C to stop)\n");

  for await (const delivery of client.listen(AGENT_ADDRESS)) {
    const intentId = delivery.intent_id;
    const status = delivery.status;
    if (intentId && ["DELIVERED", "CREATED", "IN_PROGRESS"].includes(status)) {
      console.log(`[${status}] Intent received: ${intentId}`);
      try {
        await handleIntent(client, intentId);
      } catch (e) {
        console.error(`  Error: ${e}`);
      }
    }
  }
}

main().catch(console.error);

File: java/Agent.java
/*
 * Payment service agent — Java example.
 *
 * Fetches an intent by ID, processes payment, and resumes with transaction result.
 *
 * Usage:
 *   export AXME_API_KEY="<agent-key>"
 *   javac -cp axme-sdk.jar Agent.java
 *   java -cp .:axme-sdk.jar Agent <intent_id>
 */

import dev.axme.sdk.AxmeClient;
import dev.axme.sdk.AxmeClientConfig;
import dev.axme.sdk.RequestOptions;
import java.time.Instant;
import java.util.Map;

public class Agent {
    public static void main(String[] args) throws Exception {
        if (args.length < 1) {
            System.err.println("Usage: java Agent <intent_id>");
            System.exit(1);
        }

        String apiKey = System.getenv("AXME_API_KEY");
        if (apiKey == null || apiKey.isEmpty()) {
            System.err.println("Error: AXME_API_KEY not set.");
            System.exit(1);
        }

        String intentId = args[0];
        var client = new AxmeClient(AxmeClientConfig.forCloud(apiKey));

        System.out.println("Processing intent: " + intentId);

        var intentData = client.getIntent(intentId, new RequestOptions());
        @SuppressWarnings("unchecked")
        var intent = (Map<String, Object>) intentData.getOrDefault("intent", intentData);
        @SuppressWarnings("unchecked")
        var payload = (Map<String, Object>) intent.getOrDefault("payload", Map.of());
        if (payload.containsKey("parent_payload")) {
            @SuppressWarnings("unchecked")
            var pp = (Map<String, Object>) payload.get("parent_payload");
            payload = pp;
        }

        String orderId = (String) payload.getOrDefault("order_id", "unknown");
        double amount = payload.containsKey("amount") ? ((Number) payload.get("amount")).doubleValue() : 0;
        String currency = (String) payload.getOrDefault("currency", "USD");
        String method = (String) payload.getOrDefault("method", "card");

        System.out.println("  Processing " + method + " payment: " + currency + " " + (int) amount + " for " + orderId + "...");
        Thread.sleep(1000);
        System.out.println("  Authorizing with payment provider...");
        Thread.sleep(1000);
        System.out.println("  Capturing funds...");
        Thread.sleep(1000);

        var result = Map.<String, Object>of(
            "action", "complete",
            "order_id", orderId,
            "transaction_id", "TXN-99001",
            "amount_charged", amount,
            "status", "captured",
            "processed_at", Instant.now().toString()
        );

        client.resumeIntent(intentId, result, new RequestOptions());
        System.out.println("  Payment captured: TXN-99001 (" + currency + " " + (int) amount + ")");
    }
}
